File: T3/codigos2021/T3_Elementos/app/src/main/java/com/borja/t3_elementos/adaptadores/AdaptadorRecycler.java
package com.borja.t3_elementos.adaptadores;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.borja.t3_elementos.ListasPersoActivity;
import com.borja.t3_elementos.R;
import com.borja.t3_elementos.utils.Coche;

import java.util.ArrayList;

public class AdaptadorRecycler extends RecyclerView.Adapter<AdaptadorRecycler.MiHolder> {

    ArrayList<Coche> listaDatos;
    Context context;

    public AdaptadorRecycler(ArrayList<Coche> listaDatos, Context context) {
        this.listaDatos = listaDatos;
        this.context = context;
    }

    @NonNull
    @Override
    public MiHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View vista = LayoutInflater.from(context)
                .inflate(R.layout.item_fila_recycler,parent,false);
        MiHolder holder = new MiHolder(vista);
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull MiHolder holder, int position) {

        final Coche cocheActual = listaDatos.get(position);
        //cocheActual.getMarca();
        // no meto findviews
        holder.getMarca().setText(cocheActual.getMarca());
        holder.getModelo().setText(cocheActual.getModelo());
        holder.getImagen().setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                //Toast.makeText(context,String.valueOf(cocheActual.getCv()),Toast.LENGTH_SHORT).show();
                //Intent intent = new Intent(context, ListasPersoActivity.class);
                //context.startActivity(intent);
                return true;
            }
        });
    }
    @Override
    public int getItemCount() {
        return listaDatos.size();
    }
    class MiHolder extends RecyclerView.ViewHolder{

        private TextView marca, modelo;
        private ImageView imagen;

        public MiHolder(@NonNull View itemView) {
            super(itemView);
            marca = itemView.findViewById(R.id.marca_recycler_perso);
            modelo = itemView.findViewById(R.id.modelo_recycler_perso);
            imagen = itemView.findViewById(R.id.imagen_recycler_perso);
        }

        public TextView getMarca() {
            return marca;
        }
        public TextView getModelo() {
            return modelo;
        }
        public ImageView getImagen() {
            return imagen;
        }
    }

}

File: Kotlin/T3/codigos_ret/T3_UI/app/src/main/java/com/example/t3_ui/adapters/AdaptadorRecycler.kt
package com.example.t3_ui.adapters

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.t3_ui.DetailActivity
import com.example.t3_ui.MainActivity
import com.example.t3_ui.R
import com.example.t3_ui.model.Modelo
import com.google.android.material.snackbar.Snackbar

class AdaptadorRecycler(
    var lista: ArrayList<Modelo>,
    var contexto: Context
) :
    RecyclerView.Adapter<AdaptadorRecycler.MyHolder>() {

    private lateinit var listener: OnRecyclerModeloListener

    init {
        // OnRecyclerModeloListener = SecondActivity as OnRecyclerListener
        listener = contexto as OnRecyclerModeloListener
    }


    class MyHolder(item: View) : RecyclerView.ViewHolder(item) {
        var imagen: ImageView
        var nombreModelo: TextView
        var valoracionModelo: TextView
        var boton: Button
        var botonComparar: Button;

        init {
            imagen = item.findViewById(R.id.imagen_modelo)
            nombreModelo = item.findViewById(R.id.nombre_modelo)
            valoracionModelo = item.findViewById(R.id.valoracion_modelo)
            boton = item.findViewById(R.id.boton_detalles)
            botonComparar = item.findViewById(R.id.boton_comparar)
        }
    }

    // crea el patron
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int):
            MyHolder {
        // para crear un patron necesito un view
        val vista: View = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_recycler, parent, false)
        return MyHolder(vista);
    }

    // asocia elementos en una posicion al patron que le pasa el metodo
    // anterior
    override fun onBindViewHolder(
        holder: MyHolder,
        position: Int
    ) {

        val item = lista[position]
        holder.imagen.setImageResource(item.imagen)
        holder.imagen.setOnLongClickListener {
            lista.remove(item)
            //notifyDataSetChanged()
            return@setOnLongClickListener true
        }
        holder.nombreModelo.text = item.nombre
        holder.nombreModelo.setOnClickListener {
            notifyItemMoved(position, position - 1)
        }
        holder.valoracionModelo.text = item.valoracion.toString()
        holder.boton.setOnClickListener {
            //Snackbar.make(holder.boton, item.marca, Snackbar.LENGTH_SHORT).show()
            val intent = Intent(contexto, DetailActivity::class.java)
            intent.putExtra("modelo", item)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            contexto.startActivity(intent)
        }
        holder.botonComparar.setOnClickListener {
            // disparar la acción
            listener.onModeloSelected(item)
        }


    }

    // cuantos elementos tienes que pintar en una lista
    override fun getItemCount(): Int {
        return lista.size
    }


    interface OnRecyclerModeloListener {
        fun onModeloSelected(modelo: Modelo)
    }

}